### PR TITLE
Issue 5854: Attempt to wait for Android window before reinitializing the engine

### LINKED ISSUE #5854: Crash in ANativeWindow_setBuffersGeometry
```
backtrace:
  #00  pc 000000000001b060  /system/lib64/libandroid.so (ANativeWindow_setBuffersGeometry+16)
  #00  pc 0000000000233fec  REDACTED (init_gl+328)
  #00  pc 0000000000235cf0  REDACTED (_glfwPlatformOpenWindow+76)
  #00  pc 00000000002309d8  REDACTED (glfwOpenWindow+1012)
  #00  pc 000000000015f674  REDACTED (dmGraphics::OpenGLOpenWindow(dmGraphics::Context*, dmGraphics::WindowParams*)+168)
  #00  pc 00000000000b19e8  REDACTED (dmEngine::Init(dmEngine::Engine*, int, char**)+2172)
  #00  pc 00000000000b4624  REDACTED (dmEngineCreate(int, char**)+68)
  #00  pc 00000000000b0a30  REDACTED (dmEngine::RunLoop(dmEngine::RunLoopParams const*)+72)
  #00  pc 00000000000b09c8  REDACTED (EngineMainThread(void*)+124)
  #00  pc 0000000000215ee8  REDACTED (dmThread::ThreadStartProxy(void*)+32)
  #00  pc 00000000000e205c  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+36)
  #00  pc 0000000000084af0  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64)
```

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: engine/engine/src/engine_main.cpp [hunk 1]
```diff
@@ -69,20 +69,8 @@ static void EngineMainThread(void* ctx)
     args->m_Finished = 1;
 }
 
-int engine_main(int argc, char *argv[])
+static void WaitForWindow()
```
Comment (JCash) [inline engine/engine/src/engine_main.cpp:72]: Just a small cleanup, moving code into a function

File: engine/glfw/lib/android/android_util.c [hunk 2]
```diff
@@ -123,10 +124,39 @@ static EGLint choose_egl_config(EGLDisplay display, EGLConfig* config)
     return result;
 }
 
+static int IsAppAndWindowReady(_GLFWwin_android* win)
+{
+    return win != 0 && win->app != 0 && win->app->window != 0;
```
Comment (JCash) [inline engine/glfw/lib/android/android_util.c:129]: Our working theory is that the window is 0 for some reason, and that there's a race condition between the looper and game threads during `sys.reboot()`
I.e we think the game thread restarts the engine, and tries to open a window before the OS has assigned a new window to the app.

So, let's wait for a bit, to see if that helps!

(We haven't been able to reproduce it locally)

File: engine/glfw/lib/android/android_util.c [hunk 3]
```diff
@@ -123,10 +124,39 @@ static EGLint choose_egl_config(EGLDisplay display, EGLConfig* config)
...
+}
+
 int init_gl(_GLFWwin_android* win)
 {
     LOGV("init_gl");
 
+    if (!WaitForAppAndWindow(win))
+    {
+        LOGE("ENGINE THREAD: Window not ready. Returning from init_gl()");
+        return 0;
```
Comment (JCash) [inline engine/glfw/lib/android/android_util.c:157]: If we still don't have a window, we don't want it to continue (it's what currently generates the crashes).

Instead we return 0 here, which should exit the game.
I think this is preferable to a crash.

File: engine/glfw/lib/android/android_util.c [hunk 4]
```diff
@@ -123,10 +124,39 @@ static EGLint choose_egl_config(EGLDisplay display, EGLConfig* config)
...
 
+static int IsAppAndWindowReady(_GLFWwin_android* win)
+{
+    return win != 0 && win->app != 0 && win->app->window != 0;
+}
+
+static int WaitForAppAndWindow(_GLFWwin_android* win)
+{
+    useconds_t  wait_period = 50*1000;
+    int         num_waits = 10;
```
Comment (JCash) [inline engine/glfw/lib/android/android_util.c:135]: Current max waiting time is 500ms.

**Comment by AGulev** [comment]
a bit better stac trace, maybe it will help 

```
*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 0 >>> com.potatojam.onet.master.connect.match.pairs <<<

backtrace:
  #00  pc 000000000001b2e8  /system/lib64/libandroid.so (ANativeWindow_setBuffersGeometry+16)
  #00  pc 00000000001f87a4  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (init_gl) (SourceCode: /home/runner/work/defold/defold/engine/glfw/build/../lib/android/android_util.c:154)
  #00  pc 00000000001fa4a8  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (_glfwPlatformOpenWindow) (SourceCode: /home/runner/work/defold/defold/engine/glfw/build/../lib/android/android_window.c:167)
  #00  pc 00000000001f3094  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (glfwOpenWindow) (SourceCode: /home/runner/work/defold/defold/engine/glfw/build/../lib/window.c:580)
  #00  pc 0000000000120a48  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (dmGraphics::OpenGLOpenWindow(dmGraphics::Context*, dmGraphics::WindowParams*)) (SourceCode: /home/runner/work/defold/defold/engine/graphics/build/../src/opengl/graphics_opengl.cpp:735)
  #00  pc 00000000000ab800  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (dmEngine::Init(dmEngine::Engine*, int, char**)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine.cpp:725)
  #00  pc 00000000000ae53c  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (dmEngineCreate(int, char**)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine.cpp:1942)
  #00  pc 00000000000aa854  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (dmEngine::RunLoop(dmEngine::RunLoopParams const*)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine_loop.cpp:54)
  #00  pc 00000000000aa7ec  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (EngineMain(int, char**)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine_main.cpp:49)
  #00  pc 00000000000aa7ec  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (EngineMainThread(void*)) (SourceCode: /home/runner/work/defold/defold/engine/engine/build/../src/engine_main.cpp:68)
  #00  pc 00000000001d72fc  /data/app/com.potatojam.onet.master.connect.match.pairs-oF64KaTw_tXhhMrsvnMOeA==/split_config.arm64_v8a.apk!lib/arm64-v8a/libOnetMaster.so (offset 0x1000) (dmThread::ThreadStartProxy(void*)) (SourceCode: /home/runner/work/defold/defold/engine/dlib/build/../src/dlib/thread.cpp:42)
  #00  pc 00000000000e5614  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+36)
  #00  pc 0000000000085458  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64)
 
```

**Review by britzl — APPROVED** [review]

**Pull request merged**